overlord/configstate/config/transaction.go

```diff
@@ -228,12 +233,70 @@ func (t *Transaction) GetPristineMaybe(instanceName, key string, result interfac
 	return nil
 }
 
-func getFromConfig(instanceName string, subkeys []string, pos int, config map[string]*json.RawMessage, result interface{}) error {
+// GetNoVirtual gets the configuration but ignores any virtual configuration.
+func (t *Transaction) GetNoVirtual(snapName, key string, result interface{}) error {
+	return t.get(snapName, key, result, &options{ExcludeVirtual: true})
+}
+
+type options struct {
+	ExcludeVirtual bool
+}
+
+func (t *Transaction) get(snapName, key string, result interface{}, opts *options) error {
+	t.mu.Lock()
+	defer t.mu.Unlock()
+
+	if opts == nil {
+		opts = &options{}
+	}
+
+	subkeys, err := ParseKey(key)
+	if err != nil {
+		return err
+	}
+
+	// first check if this is a virtual configuration key
+	if !opts.ExcludeVirtual {
+		virtualMu.Lock()
+		km, ok := virtualMap[snapName]
+		virtualMu.Unlock()
+		if ok && km != nil {
+			// check if this is a subkey of a virtualed key
+			for i := 0; i < len(subkeys); i++ {
+				k := strings.Join(subkeys[:len(subkeys)-i], ".")
+				if hif, ok := km[k]; ok {
+					return hif(snapName, key, VirtualResult{result})
+				}
+			}
+
+		}
+	}
+
+	// commit changes onto a copy of pristine configuration, so that get has a complete view of the config.
+	config := t.copyPristine(snapName)
+	applyChanges(config, t.changes[snapName])
+
+	purgeNulls(config)
+	return getFromConfig(snapName, subkeys, 0, config, result, opts)
+}
+
+func getFromConfig(instanceName string, subkeys []string, pos int, config map[string]*json.RawMessage, result interface{}, opts *options) error {
+	if opts == nil {
+		opts = &options{}
+	}
+
 	// special case - get root document
 	if len(subkeys) == 0 {
```

Comment:
it's a bit unclear why we would merge the virtuals only if we get the root document? we need to patch them also into subdocuments.

Reply:
That was indeed a silly oversight, I added this to the wrong place, it's fixed now